Adds auto & branches

redPlover opened the pull request:
- Includes 5-piece auto
- Contains path branches for easier creation and integration of more autos

redPlover commented:
@minhnguyenbhs I was thinking that we could use [this](https://pathplanner.dev/pplib-override-target-rotation.html) to override the rotation target. Hopefully, this could fix the other issues we were having.

minhnguyenbhs commented:
can you run this in advantagescope for me? i can't figure out how to get it to connect to the simulator

jkleiber commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -33,6 +35,7 @@ public Robot() {
 
     @Override
     public void robotInit() {
+        Pathfinding.setPathfinder(new LocalADStar());
```
Could this be done in RobotContainer instead?

jkleiber commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -200,8 +207,9 @@ public void testInit(String choice) {
     }
 
     private Translation2d getFieldToSpeaker() {
+        // return FieldConstants.fieldToRedSpeaker;
```
Could we delete this dead code?

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -65,6 +66,7 @@ public CommandSwerveDrivetrain(
         }
 
         thetaController.enableContinuousInput(-180, 180);
+        // PPHolonomicDriveController.setRotationTargetOverride(() -> calculateDesiredHeading());
```
Could we delete this dead code?

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -216,6 +230,25 @@ private Rotation2d calculateDesiredHeading(Pose2d current, Pose2d target) {
         return angle.minus(new Rotation2d(phi));
     }
 
+    /*private Optional<Rotation2d> calculateDesiredHeading() {
+
+        Pose2d current = getFieldToRobot.get();
+        Pose2d target = new Pose2d(getFieldToSpeaker.get(), new Rotation2d());
+
+        if (aligning) {
+            Pose2d robotToTarget = GeomUtil.transformToPose(current.minus(target));
+
+            Rotation2d angle =
+                    Rotation2d.fromRadians(Math.atan2(robotToTarget.getY(), robotToTarget.getX()));
+
+            angle = angle.plus(Rotation2d.fromDegrees(180));
+
+            return Optional.of(angle);
+        } else {
+            return Optional.empty();
+        }
+    }*/
```
Do we need this / can we delete it since it is commented out?

jkleiber reviewed (commented):
First pass review. It would be nice to see evidence that this works (if it does work)

minhnguyenbhs commented:
sorry i think i fixed it now- i got really confused by the new merge edits from main- i think it's working again but the tuning is really bad; how would you tune drivetrain stuff in simulation?

redPlover commented:
It seems to be an issue with auto alignment, not auto. After choosing manual alignment in teleop, the robot starts spinning in place. I'll investigate further.

redPlover commented:
@minhnguyenbhs I found the issue. The new TunerConstants were not behaving very well with our AutoAlign gains. I fixed this and now 5-note works! I also renamed your path branches so that they make sense when we flip which side we are on.

https://github.com/team401/2024-Robot-Code/assets/100500945/d63d1ee6-eb3d-48cc-a792-908b5c58bea8

jkleiber commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -90,28 +91,27 @@ public static final class FieldConstants {
+        public static final Translation2d fieldToBlueSpeaker =
+                new Translation2d(Units.inchesToMeters(-1.5), Units.inchesToMeters(218.42));
+
+        /*public static Translation2d speakerPose =
+        DriverStation.getAlliance().isEmpty()
+                        || DriverStation.getAlliance().get() == DriverStation.Alliance.Blue
+                ? new Translation2d(
+                        Units.inchesToMeters(-1.5), Units.inchesToMeters(218.42))
+                : new Translation2d(
+                        Units.inchesToMeters(652.73),
+                        Units.inchesToMeters(218.42)); // TODO: Might have to change these*/
```
Can we just delete this?

redPlover replied:
Yes, now deleted

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -133,20 +136,35 @@ private void configurePathPlanner() {
                         new ReplanningConfig()),
-                () -> false, // Change this if the path needs to be flipped on red vs blue
+                () -> {
+                    // Boolean supplier that controls when the path will be mirrored for the red
+                    // alliance
+                    // This will flip the path being followed to the red side of the field.
+                    // THE ORIGIN WILL REMAIN ON THE BLUE SIDE
```
Is this comment still accurate?

redPlover replied:
Yes. False is blue and true is red

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -205,21 +223,22 @@ private void controlDrivetrain() {
                     desiredHeading = getFieldToSource.get();
                     break;
                 case NONE:
+                    break;
```
You don't actually need to add this because it will fall through into default by design. However, you don't need to remove it now that it's added (more explicit code is good). Feel free to resolve this comment once you read it

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -238,11 +257,16 @@ private void controlDrivetrain() {
+        Translation2d robotVelocityAdjusted =
+                getRobotVelocity.get().times(DriveConstants.anticipationTime);
+
+        if (Math.abs(robotVelocityAdjusted.getNorm())
```
Does getNorm not already produce a non-negative number? (i.e. can we remove the `abs`?)

redPlover replied:
You're right, that is now deleted

jkleiber commented on src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java:
```diff
@@ -238,11 +257,16 @@ private void controlDrivetrain() {
+                getRobotVelocity.get().times(DriveConstants.anticipationTime);
+
+        if (Math.abs(robotVelocityAdjusted.getNorm())
+                < DriveConstants.minimumAnticipationVelocity) {
```
This constant is 0, so are we doing anything here? Or is this just setup for the future.

Either way feel free to leave as-is, I'm just curious about what is going on here

redPlover replied:
It is setup for the future if needed. I thought that I'd have to use it but then made it work without.

jkleiber reviewed (commented):
Some more comments before we merge this

jkleiber reviewed (approved):
Auto wins championships

Approved
